### ISSUE TITLE
Pyright doesn't report a function returning None when its return value is evaluated

### ISSUE THREAD

**Issue by gh-andre** [microsoft/pyright#5672]
**Describe the bug**

A function returning `None` should be reported when its return value is evaluated (e.g. in a conditional statement). Instead, no error is reported.

**Code or Screenshots**

```python
def fn() -> None:
    pass

if fn():
    print("A")
else:
    print("B")
    
print(fn())
```
Running this code through `mypy` v1.4.1 reports both cases:
```
> mypy test.py
test.py:4: error: "fn" does not return a value  [func-returns-value]
test.py:9: error: "fn" does not return a value  [func-returns-value]
```
Running `pyright` v1.1.320 shows no errors:
```
> pyright test.py
0 errors, 0 warnings, 0 informations
```
Python 3.10.10 on Windows

**Comment by erictraut** [comment]
This is [by design](https://microsoft.github.io/pyright/#/mypy-comparison?id=none-return-type). The `pass` statement implicitly returns `None`, so there is no type violation here. The same is true with a function that has a `return` statement with no expression or a function that "falls through" to the bottom, like these:

```python
def fn1() -> None:
    return

def fn2() -> None:
    print('')
```

The fact that mypy emits an error in this case is highly controversial and is not justified by type rules, since there is no type violation. The reasons that mypy maintainers give for emitting an error in this case are not well motivated in my opinion (and the opinions of many others). This is, at best, a code style rule that should be handled by a linter, not a static type checker.

**Comment by gh-andre** [comment]
@erictraut I used `pass` statement for brevity in the example. In the actual code the bug wasn't noticed because there was no `return` statement, but the method was called in a conditional expression.

I have to disagree. In any other language trying to evaluate a function returning `void` is clearly a bug. Evaluating `None` as a return type is not the same as evaluating a set of returned values, one of which can be `None`, which would be the case if a function returned an instance or `None`.

**Comment by erictraut** [comment]
As I mentioned, there's no basis for generating an error here from the perspective of type checking. If you decide that you want to disallow this for code style reasons, then you should use a code linter that supports such a check. Pyright is a type checker, and it reports type violations.

I understand your argument about `void`, but Python doesn't have a `void` type. The `None` type is not the same as `void`. If Python had a true `void` as other languages do, then I would agree with you. TypeScript (which is a superset of JavaScript) has both a `void` and an `undefined` in its type system. The `None` type in Python is closer in meaning to `undefined` than `void`. TypeScript generates an error if you use `void` in a conditional test, but it doesn't generate an error if you use `undefined`. Since Python has only `None`, a type checker cannot determine the intent of the programmer in this case.
